crates/flux-refineck/src/checker.rs

```diff
@@ -944,7 +949,25 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
                 Ok(Ty::tuple(tys))
             }
             Rvalue::Aggregate(AggregateKind::Closure(did, _), operands) => {
-                let upvar_tys = self.check_operands(rcx, env, stmt_span, operands)?;
+                let operand_tys = self.check_operands(rcx, env, stmt_span, operands)?;
+                let mut infcx = self.infcx(stmt_span);
+                let mut upvar_tys = vec![];
+                for ty in &operand_tys {
+                    let ref_ty = if let TyKind::Ptr(PtrKind::Mut(re), path) = ty.kind() {
+                        env.ptr_to_ref(
+                            rcx,
+                            &mut infcx,
+                            ConstrReason::Other,
+                            *re,
+                            path,
+                            PtrToRefBound::Identity,
```

Comment:
We probably want this to be `PtrToRefBound::Infer` because we want to infer a (possibly weaker) invariant on the location preserved by the closure.

```suggestion
                            PtrToRefBound::Infer,
```

Reply:
Also, a stylistic suggestion (you'll need to import `flux_common::iter::IterExt`)

```rust
                let upvar_tys = operand_tys.into_iter()
                    .map(|ty| {
                        if let TyKind::Ptr(PtrKind::Mut(re), path) = ty.kind() {
                            env.ptr_to_ref(
                                rcx,
                                &mut infcx,
                                ConstrReason::Other,
                                *re,
                                path,
                                PtrToRefBound::Infer,
                            )
                        } else {
                            Ok(ty.clone())
                        }
                    })
                    .try_collect_vec()
                    .with_span(stmt_span)?;
```